M365/MDO/ResendFailedMail.ps1

```diff
@@ -0,0 +1,330 @@
+. $PSScriptRoot\..\..\Shared\LoggerFunctions.ps1
+. $PSScriptRoot\..\..\Shared\OutputOverrides\Write-Host.ps1
+. $PSScriptRoot\..\..\Shared\OutputOverrides\Write-Progress.ps1
+. $PSScriptRoot\..\..\Shared\OutputOverrides\Write-Verbose.ps1
+. $PSScriptRoot\..\..\Shared\OutputOverrides\Write-Warning.ps1
+
+$recipientCache = @{}
+
+function Write-DebugLog ($message) {
+    if (![string]::IsNullOrEmpty($message)) {
+        $Script:DebugLogger = $Script:DebugLogger | Write-LoggerInstance $message
+    }
+}
+
+function Write-HostLog ($message) {
+    if (![string]::IsNullOrEmpty($message)) {
+        $Script:HostLogger = $Script:HostLogger | Write-LoggerInstance $message
+    }
+    # all write-host should be logged in the debug log as well.
+    Write-DebugLog $message
+}
+
+$LogFileName = "ResendFailedMail"
+$StartDate = Get-Date
+$StartDateFormatted = ($StartDate).ToString("yyyyMMddhhmmss")
+$Script:DebugLogger = Get-NewLoggerInstance -LogName "$LogFileName-Debug-$StartDateFormatted" -LogDirectory $PSScriptRoot -AppendDateTimeToFileName $false -ErrorAction SilentlyContinue
+$Script:HostLogger = Get-NewLoggerInstance -LogName "$LogFileName-Results-$StartDateFormatted" -LogDirectory $PSScriptRoot -AppendDateTimeToFileName $false -ErrorAction SilentlyContinue
+SetWriteHostAction ${Function:Write-HostLog}
+SetWriteProgressAction ${Function:Write-DebugLog}
+SetWriteVerboseAction ${Function:Write-DebugLog}
+SetWriteWarningAction ${Function:Write-HostLog}
+
+$BuildVersion = ""
+
+Write-Host ("ResendFailedMail.ps1 script version $($BuildVersion)") -ForegroundColor Green
+
+if ($ScriptUpdateOnly) {
+    switch (Test-ScriptVersion -AutoUpdate -VersionsUrl "https://aka.ms/ResendFailedMail-VersionsURL" -Confirm:$false) {
+        ($true) { Write-Host ("Script was successfully updated.") -ForegroundColor Green }
+        ($false) { Write-Host ("No update of the script performed.") -ForegroundColor Yellow }
+        default { Write-Host ("Unable to perform ScriptUpdateOnly operation.") -ForegroundColor Red }
+    }
+    return
+}
+
+if ((-not($SkipVersionCheck)) -and (Test-ScriptVersion -AutoUpdate -VersionsUrl "https://aka.ms/ResendFailedMail-VersionsURL" -Confirm:$false)) {
+    Write-Host ("Script was updated. Please re-run the command.") -ForegroundColor Yellow
+    return
+}
+
+function Test-GraphContext {
+    [OutputType([bool])]
+    param (
+        [Parameter(Mandatory = $true)]
+        [string[]]$Scopes,
+        [Parameter(Mandatory = $true)]
+        [string[]]$ExpectedScopes
+    )
+
+    $validScope = $true
+    foreach ($expectedScope in $ExpectedScopes) {
+        if ($Scopes -contains $expectedScope) {
+            Write-Verbose "Scopes $expectedScope is present."
+        } else {
+            Write-Host "The following scope is missing: $expectedScope" -ForegroundColor Red
+            $validScope = $false
+        }
+    }
+    return $validScope
+}
+
+if (-not $SkipConnectionCheck) {
```

Comment:
What is the scenario that we should support were we shouldn't validate this?

Reply:
The customer might not want or need to check the connection every time they run it in order to save time. If they run it once to see which emails would be sent, for example, why should they need to do the connection check again immediately afterwards.

Reply:
But there isn't a valid reason to assume that is done within a reasonable time or even in the same PowerShell session.

Reply:
@iserrano76 need to know what happens if the `SkipConnectionCheck` is used, when we aren't connected to EXO.